samples/matter/common/src/watchdog/watchdog.h

```diff
@@ -0,0 +1,167 @@
+	 *
+	 * You can register up to 8 Watchdog sources.
+	 *
+	 * You can feed a Watchdog source in a two ways:
+	 * 	- Manually, by calling the Feed() method in a specific context.
+	 *	- Automatically, by providing callback function and feed interval value to the constructor as an
+	 *    argument. If the callback is provided it will be called in each feed interval time periodically.
+	 */
+	class WatchdogSource {
+	public:
+		/**
+		 * @brief Callback for feeding purposes
+		 *
+		 * This callback will be run when feedingInterval timeout occurs.
+		 * It must call the Feed method within the specific context.
+		 * If this callback cannot be call within the CONFIG_NCS_SAMPLE_MATTER_WATCHDOG_TIMEOUT timeout the
+		 * reboot occurs.
+		 *
+		 * For example to feed an Watchdog source dedicated for Main thread
+		 * you can use Nrf::PostTask method:
+		 *
+		 *  Nrf::PostTask([watchdog] { watchdog->Feed(); });
+		 */
+		using FeedingCallback = void (*)(WatchdogSource *watchdogSource);
+
+		/**
+		 * @brief Create new Watchdog feeding source
+		 *
+		 * Each Watchdog source may contain a callback which is called when the feedingInterval time is elapsed.
+		 * If at least one of the watchdog source is not fed within the
+		 * CONFIG_NCS_SAMPLE_MATTER_WATCHDOG_TIMEOUT then the reboot occurs.
+		 *
+		 * @param FeedingInterval if the callback argument is provided then after elapsing time in milliseconds
+		 * of this value the callback will be set.
+		 * @param callback FeedingCallback to be called in the specific place to reset the Watchdog timer
+		 */
+		WatchdogSource(uint32_t feedingInterval = 0, FeedingCallback callback = nullptr);
+
+		/**
+		 * @brief Remove the Watchdog source
+		 *
+		 * The destructor perform a several steps to remove a WatchdogSource:
+		 * 1. Disables the Watchdog peripheral.
+		 * 2. Removes this Watchdog source.
+		 * 3. Restores all other Watchdog sources that are already registered.
+		 * 4. Enables the Watchdog peripheral.
+		 */
+		~WatchdogSource();
+
+		/**
+		 * @brief Reset the Watchdog timer from a specific source.
+		 *
+		 * If the Watchdog peripheral doesn't receive feed signal within the
+		 * CONFIG_NCS_SAMPLE_MATTER_WATCHDOG_TIMEOUT then the device reboot occurs.
+		 *
+		 * @note This method can be called within various context, wherever you want to detect high time usage,
+		 * or infinite loops. For example you can call this method within a specific thread.
+		 */
+		void Feed();
+
+		/**
+		 * @brief Restore the Watchdog source
+		 *
+		 * This method must be called in order to re-enable a Watchdog source if the Enable() static method has
+		 * been called.
+		 *
+		 * @warning This method must be called before calling Enable() if the Disable function has been called.
+		 *
+		 * @return true if a Watchdog source has been restored and started working again.
+		 * @return false if the Global Watchdog is in the state that disallows restoring the Watchdog source.
+		 */
+		bool Restore();
```

Comment:
Should this be private? This is not expected to be called by a user, right? Also, I wonder if this could be called `Install()` and reused in the constructor to reduce the code duplication.

Reply:
It can be called Install for sure :) I thought about the approach that the user wants to stop only the several watchdogs. For example, somebody creates 4 Watchog sources and wants to disable 3 of them without removing the 4th one. So it is not typical behaviour and they need to call Disable(), then Restore()/Install() on the desired sources, and then Enable(). On the other hand, after rethinking, we can do the same by calling the destructor on the required source, and thanks to that making the Install() method private.

Reply:
Ok, the rename is optional of course :). I just thought of that in the context of using it in the constructor, so installing the source for the first time.